### PR TITLE
Implement partial proofs

### PR DESCRIPTION
**Pull request by pipermerriam**
Builds on #175 

## What was wrong?

We need to be able to compute "partial" proofs.

## How was it fixed?

This PR fills out the `Proof.to_partial` object API and adds tests for it.

#### Cute Animal Picture

dogfox

### CODE REVIEW

**Comment by pipermerriam** [comment]
@marcgarreau this has been rebased.

File: ddht/v5_1/alexandria/partials/proof.py [hunk 1]
```diff
@@ -1,17 +1,23 @@
+import bisect
```
Comment (lithp) [inline ddht/v5_1/alexandria/partials/proof.py:1]: This is how you know things are about to get interesting :smirk:

**Comment by marcgarreau** [comment]
whoops, missed this one before reviewing a couple that came after. will start here this morning.

File: tests/core/v5_1/alexandria/test_get_padding_elements.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    1:             0                           P
+                 /   \                       /   \
+               /       \                   /       \
+             /           \               /           \
+    2:      0             1             0             1
+           / \           / \           / \           / \
+          /   \         /   \         /   \         /   \
+    3:   0     1       0     1       0     1       0     1
+        / \   / \     / \   / \     / \   / \     / \   / \
+    4: 0   1 0   1   0   1 0   P   0   1 0   1   0   1 0   1
```
Comment (marcgarreau) [inline tests/core/v5_1/alexandria/test_get_padding_elements.py:219]: rogue `P` at position 8

File: tests/core/v5_1/alexandria/test_get_padding_elements.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    1:             0                           1
+                 /   \                       /   \
+               /       \                   /       \
+             /           \               /           \
+    2:      0             1             0             P
+           / \           / \           / \           / \
+          /   \         /   \         /   \         /   \
+    3:   0     1       0     1       0     P       0     1
+        / \   / \     / \   / \     / \   / \     / \   / \
+    4: 0   1 0   1   0   1 0   P   0   P 0   1   0   1 0   1
```
Comment (marcgarreau) [inline tests/core/v5_1/alexandria/test_get_padding_elements.py:249]: same rogue `P`. edit: the rest of the examples have one or two of these.

File: ddht/v5_1/alexandria/partials/proof.py [hunk 4]
```diff
@@ -80,6 +164,195 @@ def get_tree(self) -> ProofTree:
...
         return self._tree_cache
 
+    def to_partial(self, start_at: int, partial_data_length: int) -> "Proof":
+        """
+        Return another proof with the minimal number of tree elements necessary
+        to prove the slice of the underlying bytestring denoted by the
+        `start_at` and `partial_data_length` parameters.
+        """
+        # First retrieve the overall data length from the proof.  The `length`
+        # shoudl always be found on the path `(True,)` which should always be present in the
```
Comment (marcgarreau) [inline ddht/v5_1/alexandria/partials/proof.py:174]: nit: `shoudl` typo and incomplete thought

**Review by marcgarreau — APPROVED** [review]
solid testing and docstrings as usual 🚀 few nits:

File: ddht/v5_1/alexandria/partials/proof.py [hunk 5]
```diff
@@ -80,6 +164,195 @@ def get_tree(self) -> ProofTree:
...
+            + padding_nodes
+        )
+        partial_elements = tuple(
+            ProofElement(path=node.path, value=node.get_value())
+            for node in partial_nodes
+        )
+        proof = Proof(self.hash_tree_root, partial_elements, self.sedes)
+        validate_proof(proof)
+
+        return Proof(self.hash_tree_root, partial_elements, self.sedes)
```
Comment (marcgarreau) [inline ddht/v5_1/alexandria/partials/proof.py:288]: whoops, apparently didnt save the last thought:
nit: many lines of comments, but still a lot going on in `to_partial`. consider decomposing into smaller funcs in a future change set.

Comment (pipermerriam) [reply ddht/v5_1/alexandria/partials/proof.py:288]: Yeah, this one is long.  I actually have a bugfix lined up for this function in a later PR.  I'll take the time to merge it back down into this one because I think I'd have caught it if I'd taken the time to split this logic out and test the individual parts.  Basically, there are a few off-by-one errors hiding in here that don't show up until you have things like partial proofs that hit certain boundary conditions that aren't currently being tested.

**Review by pipermerriam — COMMENTED** [review]
In addressing the PR feedback I ended up reworking how the proof stuff is done to eliminate the need for the `ProofTree` object entirely.  Still hacking on it, but I'll be replacing this PR with a new one that has a **lot** more granular tests and I *think* is going to be superior in terms of benchmarking.

**Comment by pipermerriam** [comment]
@marcgarreau and @lithp this now builds from #175 

The partial proof construction is now different enough that it deserves a second pass.  I was able to greatly reduce the size of `Proof.to_partial` by building on the various APIs from #175

**Comment by pipermerriam** [comment]
This has been rebased and is ready for :eyes:

File: ddht/v5_1/alexandria/partials/proof.py [hunk 6]
```diff
@@ -212,10 +305,199 @@ def path_bit_length(self) -> int:
         """
         return self.sedes.chunk_count.bit_length()  # type: ignore
 
+    def to_partial(self, start_at: int, partial_data_length: int) -> "Proof":
```
Comment (marcgarreau) [inline ddht/v5_1/alexandria/partials/proof.py:308]: nits: 
- lots of comments, but another lengthy method that might benefit from being broken down.
- I could benefit from occasional big picture context in docstrings. In this case: `to_partial` is used to send the smallest slice over the wire? To take a local proof and compare it to a received partial? both? neither?

**Review by marcgarreau — APPROVED** [review]

File: ddht/v5_1/alexandria/partials/proof.py (hunk 6 above)
Comment (marcgarreau) [reply ddht/v5_1/alexandria/partials/proof.py:308]: (we chatted. summary: `to_partial` will be used probably exclusively by the node answering a request for data in order to send the minimum data+proof footprint over the wire.)

**Comment by pipermerriam** [comment]
@lithp lemme know if you find anything you'd like me to cleanup post-merge.
